File: Platonov/2049.cpp
class Solution {
public:
    int countHighestScoreNodes(vector<int>& parents) {
        int n = parents.size();
        int* degree = new int[n];
        int* sum = new int[n];
        long long* score = new long long[n];
        for (int i = 0; i < n; ++i) {
            degree[i] = 0;
            score[i] = 1;
            sum[i] = 1;
        }
        for (int i = 0; i < n; ++i) {
            if (parents[i] != -1) {
                ++degree[parents[i]];
            }
        }
        for (int i = 0; i < n; ++i) {
            if (degree[i] == 0) {
                int v = i;
                int count = 1;
                score[i] = n - 1;
                while (parents[v] != -1) {
                    if (degree[parents[v]] == 1) {
                        sum[parents[v]] += count;
                        score[parents[v]] *= count;
                        count = sum[parents[v]];
                        if (parents[parents[v]] != -1)
                            score[parents[v]] *= (n - sum[parents[v]]);
                        v = parents[v];
                    }
                    else {
                        sum[parents[v]] += count;
                        --degree[parents[v]];
                        score[parents[v]] *= count;
                        break;
                    }
                }
            }
        }
        long long maxScore = 1;
        int countMaxScoreNodes = 0;
        for (int i = 0; i < n; ++i) {
            if (maxScore < score[i]) {
                maxScore = score[i];
                countMaxScoreNodes = 1;
            } else if (score[i] == maxScore) {
                ++countMaxScoreNodes;
            }
        }
        return countMaxScoreNodes;
    }
};

File: Чередников/medium 2049/2049.rs
fn dfs(
    graph: &Vec<Vec<usize>>,
    node: &usize,
    max_score: &mut usize,
    max_count: &mut usize,
    all: usize,
) -> usize {
    let mut count: [usize; 2] = [0; 2];
    for (idx, i) in graph[*node].iter().enumerate() {
        count[idx] = dfs(graph, i, max_score, max_count, all);
    }
    let ans = count[0] + count[1] + 1;
    let other = all - ans;

    count[0] = if count[0] == 0 { 1 } else { count[0] };
    count[1] = if count[1] == 0 { 1 } else { count[1] };
    let other = if other == 0 { 1 } else { other };

    if count[0] * count[1] * other > *max_score {
        *max_score = count[0] * count[1] * other;
        *max_count = 1;
    } else if count[0] * count[1] * other == *max_score {
        *max_count += 1;
    }
    return ans;
}

impl Solution {
    pub fn count_highest_score_nodes(parents: Vec<i32>) -> i32 {
        let mut graph = Vec::with_capacity(parents.len());
        for _ in 0..parents.len() {
            graph.push(Vec::with_capacity(2));
        }
        for (idx, el) in parents[1..].iter().enumerate() {
            let el = *el as usize;
            graph[el].push(idx+1)
        }

        let mut tmp = 0;
        let mut ans = 0;
        dfs(&graph, &0, &mut tmp, &mut ans, parents.len());
        ans as i32
    }
}